### DISCUSSION TITLE
Application Performance Monitoring view

### DISCUSSION THREAD

**Discussion by andreasgerstmayr** [perses/perses#2290] [Ideas]
I'd like to integrate an **Application Performance Monitoring** view in Perses.

In a first step, I want to show RED (rate, error and duration) metrics from traces, aggregated per service. It'll be similar to the Service Performance Monitoring tab in Jaeger UI:

![https://www.jaegertracing.io/docs/1.61/spm/](https://www.jaegertracing.io/img/frontend-ui/spm.png)

Later, I'd like to create a new overview page ("service catalog") showing all services, and indicating which services throw errors, etc. Ideally sysadmins can go to this overview page, see services which are faulty, continue to the service detail page (like the screenshot above), and from there continue to the trace (Gantt chart) and in the future potentially also to the logs of this degraded service/operation.

From a technical side, the [Span Metrics Connector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/connector/spanmetricsconnector/README.md) generates metrics from traces, which I'd like to surface in RED graphs (using the Perses TimeSeries panel).

Some parts could be a regular dashboard (the service name could be a variable), but other parts cannot be expressed in a dashboard configuration, for example the list of operations ("Operations metrics under ...") in the Jaeger UI screenshot, so it has to be a React page. _Alternatively, maybe the bottom part (the list of operations) could be a Panel? Not sure if this makes sense._

@jgbernalp suggested to implement this as a plugin, which itself embeds other panel plugins (a plugin of plugins). We can also display this APM view as a tab in the Explore/Traces page, similar to the new Finder tab in Explore/Metrics.

In the broader sense, we could call this "dynamic dashboard", i.e. a dashboard which can be configured dynamically via code instead of static JSON/YAML/Cue.

Any thoughts on this idea?

**Comment by 7840vz** [comment]
Could this be done as a Dashboard? 
```
TSpanel TSpanel TSpanel
---------
TABLE
```
?

**Reply by AntoineThebaud** [reply]
I'd also like that we manage to make this available as a regular dashboard, as it really looks like one. The main feature lacking today to achieve that are the tiny graphs embedded in the "Operations metrics under payment", which could indeed be a display option from the regular Table panel ('means that contrary to the current behavior dealing only with instant query results, it should deal with range queries here).

**Reply by andreasgerstmayr** [reply]
I agree, it looks like a dashboard. However, I don't see how we can fit the "Operations metrics under payment" in the generic Table panel we have.

The table needs the following queries:

Duration:
```
sum(
  rate(duration_milliseconds_sum{service_name="$service"}[$__rate_interval])
/
  rate(duration_milliseconds_count{service_name="$service"}[$__rate_interval])
) by (span_name)
```

Request rate
```
sum(rate(calls_total{service_name="$service"}[$__rate_interval])) by (span_name)
```

Error rate:
```
sum(rate(calls_total{service_name="$service", status_code="STATUS_CODE_ERROR"}[$__rate_interval])) by (span_name)
```

The results of these queries must be grouped by `span_name` in rows and the cells must contain a new panel (to be done) showing only the data of the specific row (`span_name`). How can we express these requirements in a declarative way?

Alternatively we could create a new panel which shows the "Operations metrics under payment" table and must have exactly the three queries from above attached. But I feel like that'd result in lots of very specific panels, vs. having general-purpose panels as we have at the moment.

**Reply by andreasgerstmayr** [reply]
Another option are repeated panels (repeated rows in this case), like here: https://grafana.com/grafana/dashboards/21202-service-performance-monitoring/

However, from an UX view, I think the table looks better (for example it contains a search box and pagination). Also, I'd expect more interactivity to come in the future.

**Reply by jgbernalp** [reply]
Pragmatic solution:
Could we also try to consider a service like a series and use only a panel group with 4 charts for (Latency, Request rate, error rate and impact)? it seems they are based on the same time range and are just separated in independent charts, but having it with series we could still convey the same information and use filtering to check only one service. It might get crowded if we expect high service cardinality.

Extended solution:
I see great value to have a table with rows that can display other plugin charts, for example in the case of troubleshooting to display nested resources in a single dashboard. In this sense, we could define variables for each row based on other values from a main query and use whatever plugin was defined for the column and resolve the "global" variables plus the "row variables". This will be probably a new panel plugin, or adjust the current table to support plugins for columns instead of single values.

**Reply by AntoineThebaud** [reply]
> The results of these queries must be grouped by span_name in rows and the cells must contain a new panel (to be done) showing only the data of the specific row (span_name). How can we express these requirements in a declarative way?

The grouping you describe is something you can do on Grafana and that we plan to bring to Perses in the coming weeks/months: it consists in having N queries defined in your panel (one for each metric), then have a merge mechanism so that the series issued from the different queries that have the same label values are merged on a single line, so then you have 1 column for each metric.

E.g something like:
![image](https://github.com/user-attachments/assets/eb845766-1c55-4405-86a3-eb74451c996f)

We are close to be able to do the same in Perses. So then what would be left missing as I was mentioning is the ability to manage range queries results (= display graphs) in cells and not only instant queries (= display instant percentage/number).

**Reply by andreasgerstmayr** [reply]
> Pragmatic solution: Could we also try to consider a service like a series and use only a panel group with 4 charts for (Latency, Request rate, error rate and impact)? [...]  It might get crowded if we expect high service cardinality.

Yes, I think this can be a problem. In Jaeger UI a paginated table is used with 20 services per page. It also contains a search bar at the top to filter for the service.

> Extended solution: In this sense, we could define variables for each row based on other values from a main query and use whatever plugin was defined for the column and resolve the "global" variables plus the "row variables".

> The grouping you describe is something you can do on Grafana and that we plan to bring to Perses in the coming weeks/months

Are you referring to the `Merge series/tables` transformation in Grafana? How are we planning to implement this in Perses, also with transformations on the panel level? Or settings specifically for the table panel?

> We are close to be able to do the same in Perses. So then what would be left missing as I was mentioning is the ability to manage range queries results (= display graphs) in cells and not only instant queries (= display instant percentage/number).

+1

Maybe this is a bit off topic now, but if anyone has ideas how to visualize the same information (RED metrics grouped by service and grouped by service/operation) in another innovative way, please let me know. The Jaeger UI screenshot above is just a starting point, we don't have to make it similar/identical.

**Comment by andreasgerstmayr** [comment]
Coming back to this topic, here's my current draft:

### Dashboard
![image](https://github.com/user-attachments/assets/fa201e8a-5bb8-4865-a2fd-03b54cd54da9)

### Column Settings
![image](https://github.com/user-attachments/assets/fc400a1e-1a48-4f30-b009-254c3b40c8c7)

### Panel JSON

<details>
<summary>Show</summary>

```
{
  "kind": "Panel",
  "spec": {
    "display": {
      "name": "Operation Metrics"
    },
    "plugin": {
      "kind": "Table",
      "spec": {
        "cellSettings": [],
        "columnSettings": [
          {
            "header": "Name",
            "name": "span_name"
          },
          {
            "cellType": "sparkline",
            "header": "Request rate",
            "name": "value #1"
          },
          {
            "cellType": "sparkline",
            "header": "Error rate",
            "name": "value #2"
          },
          {
            "cellType": "sparkline",
            "header": "Duration",
            "name": "value #3"
          }
        ],
        "defaultColumnHeight": 90,
        "defaultColumnWidth": "auto",
        "density": "standard",
        "transforms": [
          {
            "kind": "MergeSeries",
            "spec": {}
          }
        ]
      }
    },
    "queries": [
      {
        "kind": "TimeSeriesQuery",
        "spec": {
          "plugin": {
            "kind": "PrometheusTimeSeriesQuery",
            "spec": {
              "datasource": {
                "kind": "PrometheusDatasource",
                "name": "Prometheus"
              },
              "query": "sum(rate(traces_span_metrics_calls{service_name=\"$service\"}[$__rate_interval])) by (span_name) > 0"
            }
          }
        }
      },
      {
        "kind": "TimeSeriesQuery",
        "spec": {
          "plugin": {
            "kind": "PrometheusTimeSeriesQuery",
            "spec": {
              "datasource": {
                "kind": "PrometheusDatasource",
                "name": "Prometheus"
              },
              "query": "sum(rate(traces_span_metrics_calls{service_name=\"$service\", status_code=\"STATUS_CODE_ERROR\"}[$__rate_interval])) by (span_name) > 0"
            }
          }
        }
      },
      {
        "kind": "TimeSeriesQuery",
        "spec": {
          "plugin": {
            "kind": "PrometheusTimeSeriesQuery",
            "spec": {
              "datasource": {
                "kind": "PrometheusDatasource",
                "name": "Prometheus"
              },
              "query": "sum(\n  rate(traces_span_metrics_duration_sum{service_name=\"$service\"}[5m])\n  /\n  rate(traces_span_metrics_duration_count{service_name=\"$service\"}[5m])\n) by (span_name) > 0"
            }
          }
        }
      }
    ]
  }
}
```

</details>

# Questions
* How should we support setting panel options of the embedded panel (in this case the `StatChart` panel)? For example, I want to set the units of the `StatChart` panel for each column.
* How can I improve the UI/UX, similar to Jaeger UI (screenshot above)?
  * search box to search by operation name - how should it be generic for the table panel?
  * supporting clicking on the table rows
  * sorting - while the table panel supports sorting for the case where each cell is a single string or number, how would it work in this case, where each cell is a timeseries? Not sure how maintainable the Table panel will be when we add every "exotic" use case like this :thinking:
